lib/puppet/functions/ensure_at_least.rb

```diff
@@ -0,0 +1,72 @@
+# @summary
+#   Return the package version for a given package that is at least the given minimum version
+#   The function will compare the currently installed package with the given minimum version and:
+#    - return the minimum version if the installed package is less than the minimum version, or not present
+#    - return the installed version if the installed package is the same or newer than the minimum version
+#
+# @example Example usage:
+#   # Ensure at least version 2.7.5-77.el7_6 of python is installed
+#     package { 'python':
+#       ensure => ensure_at_least('python', '2.7.5-77.el7_6')
+#     }
+#
+# @example To specify a specific provider:
+#   # Ensure at least version 2.7.5-77.el7_6 of python is installed
+#     package { 'openssl':
+#       ensure   => ensure_at_least('openssl', '2.1.2', 'gem'),
+#       provider => gem
+#     }
+#
+Puppet::Functions.create_function(:ensure_at_least) do
```

Comment:
We should namespace this function to `stdlib::ensure_at_least`.

Most of the functions in stdlib aren't namespaced because they predate the capability; namespacing functions wasn't possible until Puppet 4.

Some functions arguably made sense without a stdlib namespace. Basic string manipulations, comparisons... things that should probably just be part of core Puppet really. E.g. `join()`.

This function is a niche use case that's effectively working around a deficiency in Puppet's ability to express desired state of a Package. It wouldn't make sense to have this in Puppet core, because if we were going to put something in core it would be more integrated into the Package type instead. As such, while this is a powerful tool and useful perhaps to have in Stdlib, I think it makes sense to namespace it so that it is clear where it comes from and that it is an add-on functionality.

Reply:
Following that line of reasoning, it could also be delivered as a separate module on the Forge and not as part of stdlib. It remains debatable if it should still be namespaced yes or no in that case, as a separate module would ensure that it is seen as add-on functionality and not core. I personally would like to see as short a name as possible for the function, which namespacing wouldn't help with. I don't we can use `init.rb` for a function to give it the name of the module, right?

Reply:
Not init.rb, no, but from a technical standpoint, a namesake function can be defined in a module as e.g.

```
ensure_at_least/lib/puppet/functions/ensure_at_least.rb
```

That said, I think a longer name in stdlib would have a higher likelihood of adoption and better maintenance. Because of the specificity of the use case I don't agree that a short name is particularly desirable. As such, I would favor the longer name in stdlib over a shorter name in a standalone module for this function specifically.

As I said before, I agree that short names for some functions _is_ desirable – just not for this one.

Reply:
Agreed, I'll move it to `stdlib::ensure_at_least`